mne/_fiff/meas_info.py

```diff
@@ -1669,43 +1775,26 @@ class Info(ValidatedDict, SetChannelsMixin, MontageMixin, ContainsMixin):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self._unlocked = True
-        # Deal with h5io writing things as dict
-        if "bads" in self:
-            self["bads"] = MNEBadsList(bads=self["bads"], info=self)
-        for key in ("dev_head_t", "ctf_head_t", "dev_ctf_t"):
-            _format_trans(self, key)
-        for res in self.get("hpi_results", []):
-            _format_trans(res, "coord_trans")
-        if self.get("dig", None) is not None and len(self["dig"]):
-            if isinstance(self["dig"], dict):  # needs to be unpacked
-                self["dig"] = _dict_unpack(self["dig"], _DIG_CAST)
-            if not isinstance(self["dig"][0], DigPoint):
-                self["dig"] = _format_dig_points(self["dig"])
-        if isinstance(self.get("chs", None), dict):
-            self["chs"]["ch_name"] = [
-                str(x) for x in np.char.decode(self["chs"]["ch_name"], encoding="utf8")
-            ]
-            self["chs"] = _dict_unpack(self["chs"], _CH_CAST)
-        for pi, proj in enumerate(self.get("projs", [])):
-            if not isinstance(proj, Projection):
-                self["projs"][pi] = Projection(**proj)
-        # Old files could have meas_date as tuple instead of datetime
-        try:
-            meas_date = self["meas_date"]
-        except KeyError:
-            pass
-        else:
-            self["meas_date"] = _ensure_meas_date_none_or_dt(meas_date)
+        # Deal with h5io writing things as dict and restore MNE-specific types
+        _restore_mne_types(self)
         self._unlocked = False
-        # with validation and casting
-        for key in ("helium_info", "subject_info"):
-            if key in self:
-                self[key] = self[key]
 
     def __setstate__(self, state):
-        """Set state (for pickling)."""
+        """Set state (for unpickling and JSON deserialization)."""
+        # Call parent __setstate__ which will call _restore_objects
+        # to convert JSON primitives and transform dicts to native types
         super().__setstate__(state)
-        self["bads"] = MNEBadsList(bads=self["bads"], info=self)
+
+        # Restore MNE-specific types (requires unlocked state)
+        # This reconstructs MNEBadsList, DigPoint, Projection, etc.
+        was_locked = not self._unlocked
+        if was_locked:
+            self._unlocked = True
+
+        _restore_mne_types(self)
+
+        if was_locked:
+            self._unlocked = False
```

Comment:
Why not
```suggestion
        with self._unlock():
            _restore_mne_types(self)
```
I think it should behave the same way. You are basically ensuring `self._unlocked = True` then setting `self._unlocked` to the state it was in prior to that, which is what `self._unlock()` context manager does